falcon/asgi/response.py

```diff
@@ -0,0 +1,256 @@
+# Copyright 2019 by Kurt Griffiths
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""ASGI Response class."""
+
+from inspect import iscoroutine
+
+import falcon.response
+
+__all__ = ['Response']
+
+
+class Response(falcon.response.Response):
+    """
+
+    Attributes:
+        sse (coroutine): A Server-Sent Event (SSE) emitter, implemented as
+            an async coroutine function that returns an iterable
+            of :py:class:`falcon.asgi.SSEent` instances. Each event will be
+            serialized and sent to the client as HTML5 Server-Sent Events.
+
+        data (bytes): Byte string representing response content.
+
+            Use this attribute in lieu of `body` when your content is
+            already a byte string (of type ``bytes``).
+
+            Warning:
+                Always use the `body` attribute for text, or encode it
+                first to ``bytes`` when using the `data` attribute, to
+                ensure Unicode characters are properly encoded in the
+                HTTP response.
+
+            Note:
+                Unlike the WSGI Response class, the ASGI Response class
+                does not implement the side-effect of serializing
+                the media object (if one is set) when the `data`
+                attribute is read. Instead,
+                :py:meth:`falcon.asgi.Response.render_body` should
+                be used to get the correct content for the response.
+    """
+
+    # PERF(kgriffs): These will be shadowed when set on an instance; let's
+    #   us avoid having to implement __init__ and incur the overhead of
+    #   an additional function call.
+    _sse = None
+    _registered_callbacks = None
+    _media_rendered = None
+
+    @property
+    def sse(self):
+        return self._sse
+
+    @sse.setter
+    def sse(self, value):
+        self._sse = value
+
+    @property
+    def media(self):
+        return self._media
+
+    @media.setter
+    def media(self, value):
+        self._media = value
+        self._media_rendered = None
+
+    @property
+    def data(self):
```

Comment:
We'll have to also document the differences between this property and the WSGI `Request.data`, i.e. that accessing this property does not "materialize" `resp.media`, and one should use `render_body` instead. This bit me while trying to implement the async flavour of the "Look" tutorial :slightly_smiling_face:

Reply:
Do you think we should change this on the WSGI side to be the same as for ASGI, or just leave as-is with some warnings in the docs?

Reply:
IMHO it is fine to just document this for now (the first alpha) as we will probably have to make slight tweaks based on community feedback anyway.